### ISSUE TITLE
Return type not inferred in fit(Histogram) after v0.13.1 => v0.14.0

### ISSUE THREAD

**Issue by sbromberger** [JuliaStats/StatsBase.jl#252]
Minimal example:
```
julia> using StatsBase, Base.Test

julia> v = rand(1:20, 200);

julia> @inferred(fit(Histogram, v))
WARNING: Default for keyword argument "closed" has changed from :right to :left.
Stacktrace:
 [1] depwarn(::String, ::Symbol) at ./deprecated.jl:64
 [2] _check_closed_arg at /Users/bromberger1/.julia/v0.6/StatsBase/src/hist.jl:11 [inlined]
 [3] #fit#66(::Symbol, ::Bool, ::Int64, ::Function, ::Type{StatsBase.Histogram{Int64,N,E} where E where N}, ::Tuple{Array{Int64,1}}) at /Users/bromberger1/.julia/v0.6/StatsBase/src/hist.jl:266
 [4] (::StatsBase.#kw##fit)(::Array{Any,1}, ::StatsBase.#fit, ::Type{StatsBase.Histogram{Int64,N,E} where E where N}, ::Tuple{Array{Int64,1}}) at ./<missing>:0
 [5] fit(::Type{StatsBase.Histogram{Int64,N,E} where E where N}, ::Array{Int64,1}) at /Users/bromberger1/.julia/v0.6/StatsBase/src/hist.jl:213
 [6] #fit#69(::Array{Any,1}, ::Function, ::Type{StatsBase.Histogram}, ::Array{Int64,1}, ::Vararg{Array{Int64,1},N} where N) at /Users/bromberger1/.julia/v0.6/StatsBase/src/hist.jl:278
 [7] fit(::Type{StatsBase.Histogram}, ::Array{Int64,1}) at /Users/bromberger1/.julia/v0.6/StatsBase/src/hist.jl:278
 [8] eval(::Module, ::Any) at ./boot.jl:235
 [9] eval_user_input(::Any, ::Base.REPL.REPLBackend) at ./REPL.jl:66
 [10] macro expansion at ./REPL.jl:97 [inlined]
 [11] (::Base.REPL.##1#2{Base.REPL.REPLBackend})() at ./event.jl:73
while loading no file, in expression starting on line 0
ERROR: return type StatsBase.Histogram{Int64,1,Tuple{StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}}}} does not match inferred return type StatsBase.Histogram{_,1,_} where _ where _
Stacktrace:
 [1] error(::String) at ./error.jl:21
```

(This worked fine in StatsBase v0.13.1.)

**Comment by andreasnoack** [comment]
@oschulz It would be great if you could take a look at this. For some reason, the return type is not tightly inferred anymore.

**Comment by sbromberger** [comment]
Thanks and sorry for the apparently hyperbolic title - the failure is in LightGraphs tests, where we use `@inferred` pretty much everywhere so we can pinpoint performance issues.

**Comment by andreasnoack** [comment]
No problem. It is great that you caught the regression with your test coverage.

**Comment by oschulz** [comment]
Thanks for discovering this, @sbromberger!

@andreasnoack, I'm on it.

**Comment by oschulz** [comment]
See #253.

**Comment by jpfairbanks** [comment]
Yay decentralized testing ecosystem! Chalk one up for network effects.

**Comment by sbromberger** [comment]
@oschulz @andreasnoack thank you very much! Could you please tag a new version so we can get LightGraphs building properly again?

**Comment by ararslan** [comment]
https://github.com/JuliaLang/METADATA.jl/pull/9115

**Comment by ararslan** [comment]
@sbromberger Can you confirm that this is fixed for you with StatsBase v0.14.1?

**Comment by sbromberger** [comment]
It's fixed. Thank you very much!

**Comment by sbromberger** [comment]
Though that warning about left/right is a bit confusing. We're using the default.

**Comment by ararslan** [comment]
Yes, it is admittedly rather confusing. The upshot is that we wanted to change the default value from right to left, but if you do that out of the blue is suddenly and subtly breaks a lot of user code. So we came up with a way that will emit a deprecation warning if you use the default--this is to encourage users to think about whether they need right- or left-closure and make that explicit in their code so that nothing breaks when we change the default.

I'll close this issue since the issue originally stated has been addressed.
